File: app/src/main/java/me/devsaki/hentoid/parsers/images/ImageListParser.java
package me.devsaki.hentoid.parsers.images;

import android.util.Pair;

import androidx.annotation.NonNull;

import com.annimon.stream.Optional;

import org.apache.commons.lang3.tuple.ImmutablePair;

import java.io.IOException;
import java.util.List;
import java.util.Map;

import me.devsaki.hentoid.database.domains.Chapter;
import me.devsaki.hentoid.database.domains.Content;
import me.devsaki.hentoid.database.domains.ImageFile;
import me.devsaki.hentoid.util.exception.EmptyResultException;
import me.devsaki.hentoid.util.exception.LimitReachedException;

public interface ImageListParser {
    List<ImageFile> parseImageList(@NonNull Content content) throws Exception;

    List<ImageFile> parseImageList(@NonNull Content onlineContent, @NonNull Content storedContent) throws Exception;

    ImmutablePair<String, Optional<String>> parseImagePage(@NonNull String url, @NonNull List<Pair<String, String>> requestHeaders) throws IOException, LimitReachedException, EmptyResultException;

    Optional<ImageFile> parseBackupUrl(
            @NonNull String url,
            @NonNull Map<String, String> requestHeaders,
            int order,
            int maxPages,
            Chapter chapter) throws Exception;
}

File: app/src/main/java/me/devsaki/hentoid/parsers/images/ImageListParser.kt
package me.devsaki.hentoid.parsers.images

import me.devsaki.hentoid.database.domains.Chapter
import me.devsaki.hentoid.database.domains.Content
import me.devsaki.hentoid.database.domains.ImageFile
import me.devsaki.hentoid.util.exception.EmptyResultException
import me.devsaki.hentoid.util.exception.LimitReachedException
import java.io.IOException

interface ImageListParser {
    /**
     * Parse image list from the given URL, enriching the given Content if relevant (see Content.updatedProperties)
     */
    @Throws(Exception::class)
    fun parseImageList(content: Content, url: String): List<ImageFile>

    /**
     * Parse image list from the given online Content, retrieving only images from chapters
     * that aren't already in the given stored Content
     */
    @Throws(Exception::class)
    fun parseImageList(onlineContent: Content, storedContent: Content): List<ImageFile>

    /**
     * Returns image URL and backup URL from the given page URL
     */
    @Throws(IOException::class, LimitReachedException::class, EmptyResultException::class)
    fun parseImagePage(
        url: String,
        requestHeaders: List<Pair<String, String>>
    ): Pair<String, String?>

    @Throws(Exception::class)
    fun parseBackupUrl(
        url: String,
        requestHeaders: Map<String, String>,
        order: Int,
        maxPages: Int,
        chapter: Chapter?
    ): ImageFile?

    fun getAltUrl(url: String): String

    fun clear()
}